### ISSUE TITLE
Cannot add Owner via json rpc calls

### ISSUE THREAD

**Issue by mr-rooftop** [icon-project/multisig-wallet#26]
I discussed this issue already on ICON Dev telegram group, but couldn't resolve it.

I want to add an Owner as described in the README with submitTransaction and confirmTransaction.

local tbears docker container via 
```
docker run -it --name local-tbears -p 9000:9000 -v ~path/to/multisig-wallet:/work iconloop/tbears:mainnet
```

Deployed the SCORE with 2 owners and required == 2.
tbears_cli_config.json:
```
{
    "uri": "http://127.0.0.1:9000/api/v3",
    "nid": "0x3",
    "from": "hxe7af5fcfd8dfc67530a01a0e403882687528dfcb",
    "to": "cx0000000000000000000000000000000000000000",
    "deploy": {
        "mode": "install",
        "contentType": "zip",
        "scoreParams": {
            "_walletOwners": "hxe7af5fcfd8dfc67530a01a0e403882687528dfcb,hx5ecf7ddc4ce809f948513b8fb1d599467c1ed5c0",
            "_required": "0x02"
        }
    },
    "txresult": {},
    "transfer": {}
}
```

Json to submit via `tbears sendtx` (taken from README):

```
{
  "jsonrpc": "2.0",
  "method": "icx_sendTransaction",
  "params": {
    "version": "0x3",
    "from": "hxe7af5fcfd8dfc67530a01a0e403882687528dfcb",
    "value": "0x0",
    "stepLimit": "0x3000000",
    "nid": "0x3",
    "nonce": "0x1",
    "to": "cxa6a4681e7e49b15715d2f5cfe5bddedacae4940c",
    "dataType": "call",
    "data": {
      "method": "submitTransaction",
      "params": {
        "_destination": "cxa6a4681e7e49b15715d2f5cfe5bddedacae4940c",
        "_method": "addWalletOwner",
        "_params": "[{\"name\":\"_walletOwner\",\"type\":\"Address\",\"value\":\"hx22cf7dd5ece809f948513b8fb1d599467c1ed5c1\"}]",
        "_description": "add new owner in wallet"
      }
    }
  },
  "id": 1
}
```

This works out.
Then the confirmation json (taken from README):
```
{
  "jsonrpc": "2.0",
  "method": "icx_sendTransaction",
  "params": {
    "version": "0x3",
    "from": "hx5ecf7ddc4ce809f948513b8fb1d599467c1ed5c0",
    "value": "0x0",
    "stepLimit": "0x30000000",
    "nid": "0x3",
    "nonce": "0x1",
    "to": "cxa6a4681e7e49b15715d2f5cfe5bddedacae4940c",
    "dataType": "call",
    "data": {
      "method": "confirmTransaction",
      "params": {
        "_transactionId": "0x0"
      }
    }
  },
  "id": 1
}
```

If I inspect now this transaction with `tbears txresult`, I get `key is none`:
```
root@33ecf80960d7:/work# tbears txresult 0x10e5f3c00d5d51bae4d8d50b2e208203aaf4021cec161a18e7f0a3bf4598fcf9
Transaction : {
    "jsonrpc": "2.0",
    "result": {
        "txHash": "0x10e5f3c00d5d51bae4d8d50b2e208203aaf4021cec161a18e7f0a3bf4598fcf9",
        "blockHeight": "0x1f6",
        "blockHash": "0xe9ac172a591e876db5027b6fdd6e3676090bae775b50bf0ba7087410e8584117",
        "txIndex": "0x0",
        "to": "cxa6a4681e7e49b15715d2f5cfe5bddedacae4940c",
        "stepUsed": "0xf9e70",
        "stepPrice": "0x0",
        "cumulativeStepUsed": "0xf9e70",
        "eventLogs": [],
        "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
        "status": "0x0",
        "failure": {
            "code": "0x6",
            "message": "key is None"
        }
    },
    "id": 1234
}
```

This is my tbears.log output:
![Screenshot_20200606_203420](https://user-images.githubusercontent.com/47558538/84006369-cd237e80-a998-11ea-9b7b-5a1863dac989.png)

It seems like a `None` value gets added during the "first commit iteration" (there are two commmit iterations, one to check the TX and one to commit it to the blockchain) to an ArrayDB, and then the "second iteration" tries to access the None value, which fails.

At some point at this line https://github.com/icon-project/multisig-wallet/blob/72201bd59bffd374930542d95fbb129a3ca9a85b/multisig_wallet/multisig_wallet.py#L248 the if gets a `None` value as wallet_owner, that's why it fails.

It seems as if this `None` gets added during the "first commit iteration", because my "debug" reverts  didn't catch anything in the first iteration (didn't fail during sendtx), but then failed when I inspected the TX with txresult.

Then I put this if check before the problematic if:
```
        for wallet_owner in self._wallet_owners:
            if wallet_owner is None:
                continue
            if self._confirmations[transaction_id][wallet_owner]:
                count += 1
```
Guess what? The TX went successfully through.
Guess even more? This is the return value if I call getWalletOwners:
![Screenshot_20200606_204616](https://user-images.githubusercontent.com/47558538/84006553-2390bd00-a999-11ea-827f-de2bbba1dff2.png)

Now I submitted the "addOwner" TX, and afterwards I have the added Address as well as `None` as an owner!

I thought that I had maybe a faulty installation of tbears or some other issue from my machine, but my colleague had the same problems if he used the same setup from his machine.

**What do I have to do so I can use the MultisigWallet ICON wallet?**

Best Regards,
Mr-Rooftop

**Comment by soobokjin** [comment]
Thank you for your interest. After I read your report, I will leave a comment here.

**Comment by mr-rooftop** [comment]
Thanks a lot!
Alternatively, you could also give me a step by step explanation of how to successfully use the ICON MultisigWallet.

**Comment by sink772** [comment]
@mr-rooftop I didn't follow your steps, however, there is a good example of how to use the multisig-wallet.  Please check the following.

https://github.com/icon-project/icon-testsuite/blob/master/test/java/foundation/icon/test/cases/MultiSigWalletTest.java

When I tested with initial 2 owner addresses, I managed to run the test above successfully.
If you compare two cases, I think you will find the root cause.

**Comment by soobokjin** [comment]
> Thanks a lot!
> Alternatively, you could also give me a step by step explanation of how to successfully use the ICON MultisigWallet.

I have tested on the same env (i.e. tbears docker v.1.6.4) and found the same issue. After check analyzes the reason, leave comments here. Thank you for the report.

**Comment by soobokjin** [comment]
@mr-rooftop 
Could you send the below query request (using 'tbears call' command) in your t-bears container environment and leave the response here?
```
{
    "jsonrpc": "2.0",
    "id": 100,
    "method": "icx_call",
    "params": {
        "to": "cx0000000000000000000000000000000000000001",
        "dataType": "call",
        "data": {
            "method": "getRevision"
        }
    }
}
```

**Comment by mr-rooftop** [comment]
@soobokjin 
I ran the json you gave me before deploying the multisigWallet and after deploying it.

I also thought that maybe in `"to": "cx0000000000000000000000000000000000000001"` somehow the `1` at the end was mistyped, replaced it with `0` and ran your json again.

In any of the above scenarios I got:
![Screenshot_20200612_115642](https://user-images.githubusercontent.com/47558538/84466777-6a85f780-aca4-11ea-9da4-0b42652084cd.png)

Thx for dealing with the issue! :pray:

**Comment by soobokjin** [comment]
I guess the reason why raising 'key is None' error is you have manipulated (cleaned) the state DB('.statedb', '.score').

Basically, when tbears docker container is run, it automatically updates the environment internally to match with mainnet env.

It occurs only once a container is created (docker run).

If you reset the state DB after creating a container or use another stateDB which is not updated, multisig could not work properly.

So, try removing your docker container, state DB (.statedb, .score) and rerun tbears container.

After re-run it, check the 'getRevision'. you can get below response.
```
root@26d104ad8b9e:/work# tbears call multi/querys/revision.json 
response : {
    "jsonrpc": "2.0",
    "result": {
        "code": "0x8",
        "name": "1.5.20"
    },
    "id": 1234
}
```

In this environment, retry 'addwalletOwner'. you can add a wallet owner successfully.

Finally, I recommend not to manipulate your state DB as the SCORE could not work properly.

Hope this solution worked well.

**Comment by mr-rooftop** [comment]
I did
```
sudo rm -rf .score .statedb rc.log tbears.log tbears_server_config.json
docker rm local-tbears
docker rmi TBEARS-IMAGE
```
and just to be super sure removed every currently unused container/image/network/volume on my machine:
```
docker system prune -a
```

Then I spun tbears up again with:
```
docker run -it --name local-tbears -p 9000:9000 -v ~path/to/multisig-wallet:/work iconloop/tbears:mainnet
```
and I got the same response as before when I ran the `getRevision` json:
```
root@92fa32b61732:/work# tbears call tbears_calls/get_revision.json 
{
    "jsonrpc": "2.0",
    "error": {
        "code": -32003,
        "message": "Method not found: Governance.getRevision"
    },
    "id": 1234
}
```

Do I have to remove something else that I'm not aware of on my system?

BTW: 
The colleague I mentioned in the initial post installed everything from a clean state, and it still failed on his machine.

**Comment by mr-rooftop** [comment]
Just cloned the repo anew on a ubuntu instance in the cloud, still same issue.

Maybe the `getRevision.json` or my `docker run` command is faulty?

Possibly you used a little bit different commands to achieve success, and I'm overlooking some detail. 
Perhaps a tbears image with a different tag than `mainnet`?

**Comment by soobokjin** [comment]
It seems to docker volume mount issue.

Could you run the docker container with below commands and check the revision? The difference is the container volume mount path.

```
docker run -it --name local-tbears -p 9000:9000 -v ~path/to/multisig-wallet:/work/multi iconloop/tbears:mainnet
```
It seems that the t-bears docker is written without considering the docker volume option.

I am going to share this issue with my team and modify it. It may take time to modify it.

Thank you for explaining the problem in detail.

**Comment by mr-rooftop** [comment]
Wow that solved it, 42 thanks to you @soobokjin ! :pray: :partying_face: 

I don't quite get what the issue with the volume is, could you explain that in a few short sentences?

No problem if not, you guys did a really great support, I'm happy that I can finally use the wonderful MultisigWallet! :smiley:

**Comment by soobokjin** [comment]
Sorry for the late response! 

First of all, I am so happy that the problem has been solved. Above all, thanks to the detailed explanation, we found the inconvenience of our service and got a chance to improve it.

The cause of the problem is described as follows.

At the time of the image being built, the stateDB is created and updated internally, in '/work' directory. 

If you use mount option (~/path-to-multisig:/work),  created stateDB on image build becomes invisible, and new stateDB is created on the host directory path (i.e. ~/path-to-multisig). The newly created DB is not updated.

That's why you got an error. 't-bears' use the newly created stateDB on your host path instead of the DB which was created on image build.

If you have any more questions, I will answer in detail.

Thank you!
